### PR TITLE
Update MSFT_IntuneAppProtectionPolicyiOS.psm1 - fixes #2942

### LINKED ISSUE #2942: IntuneAppProtectionPolicyIOS - Test-TargetResource returns false even if config matches existing policy
The Test-TargetResource function in the IntuneAppProtectionPolicyIOS returns false even though it should return true.
This can be seen either through the verbose output of the start-DSCConfiguration command or in the output of Test-DSCConfiguration. 
The Test will report a false match even using a config exported from the tenant immediately before running the configuration, so the exported document does not return a true value for the policy it was exported from

The cause seems to be that the command Get-MgDeviceAppManagementiOSManagedAppProtection in  Get-TargetResource is running in the v1.0 profile rather than the beta profile as specified at the start of that function - this means the function returns null values for some settings which are correctly listed on the exported document.

#### Verbose logs showing the problem
I've pasted the current/expected config results and the false outcome - this is using a newly created test profile
VERBOSE: [COMPUTER]:                            [[IntuneAppProtectionPolicyiOS]12345678-1234-1234-1234-123456789123]
Current Values: AllowedDataStorageLocations=(); AllowedInboundDataTransferSources=allApps;
AllowedIosDeviceModels=$null; AllowedOutboundClipboardSharingExceptionLength=$null;
AllowedOutboundClipboardSharingLevel=managedAppsWithPasteIn; AllowedOutboundDataTransferDestinations=allApps;
AppActionIfDeviceComplianceRequired=; AppActionIfIosDeviceModelNotAllowed=; AppActionIfMaximumPinRetriesExceeded=;
AppDataEncryptionType=whenDeviceLocked; ApplicationId=***; ApplicationSecret=***;
Apps=(com.microsoft.bing.halseyassistant,com.microsoft.d365.fs.mobile,com.microsoft.dynamics,com.microsoft.dynamics.inv
oice,com.microsoft.dynamics.iphone.moca,com.microsoft.dynamics.iphone.moca.fieldservices,com.microsoft.dynamics.iphone.
moca.sales,com.microsoft.lync2013.iphone,com.microsoft.mobile.polymer,com.microsoft.msapps,com.microsoft.msedge,com.mic rosoft.o365shdmobileapp,com.microsoft.office.excel,com.microsoft.office.outlook,com.microsoft.office.powerpoint,com.mic rosoft.office.word,com.microsoft.office365booker,com.microsoft.officelens,com.microsoft.officemobile,com.microsoft.onen ote,com.microsoft.plannermobile,com.microsoft.powerbimobile,com.microsoft.procsimo,com.microsoft.ramobile,com.microsoft .rms-sharing,com.microsoft.scmx,com.microsoft.sharepoint,com.microsoft.shiftr,com.microsoft.skydrive,com.microsoft.skyp e.teams,com.microsoft.splists,com.microsoft.stream,com.microsoft.to-do,com.microsoft.visio,com.microsoft.whiteboard,com .microsoft.workfolders,com.veradocs.ios.appstore.intune,wefwef); Assignments=(); CertificateThumbprint=;                ContactSyncBlocked=False; Credential=$null; CustomBrowserProtocol=; DataBackupBlocked=False; Description=some           descriptive words to describe the policy in the description. indescribable; DeviceComplianceRequired=True;              DisableAppPinIfDevicePinIsSet=False; DisableProtectionOfManagedOutboundOpenInData=$null; DisplayName=testiospolicy;     Ensure=Present; ExcludedGroups=(); ExemptedAppProtocols=(); FaceIdBlocked=False; FilterOpenInToOnlyManagedApps=$null;   FingerprintBlocked=False; Identity=T_12345678-4321-4321-4321-123456789123; ManagedBrowser=notConfigured;                ManagedBrowserToOpenLinksRequired=False; Managedidentity=False; MaximumPinRetries=5; MinimumPinLength=4;                MinimumRequiredAppVersion=$null; MinimumRequiredOsVersion=$null; MinimumRequiredSdkVersion=$null;                       MinimumWarningAppVersion=$null; MinimumWarningOSVersion=$null; MinimumWipeAppVersion=$null; MinimumWipeOSVersion=$null;  MinimumWipeSdkVersion=$null; NotificationRestriction=; OrganizationalCredentialsRequired=False;                        PeriodBeforePinReset=90.00:00:00; PeriodOfflineBeforeAccessCheck=12:00:00;                                              PeriodOfflineBeforeWipeIsEnforced=90.00:00:00; PeriodOnlineBeforeAccessCheck=00:30:00; PinCharacterSet=numeric;         PinRequired=True; PinRequiredInsteadOfBiometricTimeout=$null; PrintBlocked=False;                                       ProtectInboundDataFromUnknownSources=$null; SaveAsBlocked=False; SimplePinBlocked=True; TargetedAppManagementLevels=;   TenantId=*** 
VERBOSE: [COMPUTER]:                            [[IntuneAppProtectionPolicyiOS]12345678-1234-1234-1234-123456789123]
Target Values: AllowedDataStorageLocations=(); AllowedInboundDataTransferSources=allApps;                               AllowedOutboundClipboardSharingExceptionLength=0; AllowedOutboundClipboardSharingLevel=managedAppsWithPasteIn;          AllowedOutboundDataTransferDestinations=allApps; AppActionIfDeviceComplianceRequired=block;                             AppActionIfIosDeviceModelNotAllowed=block; AppActionIfMaximumPinRetriesExceeded=block;                                  AppDataEncryptionType=whenDeviceLocked; ApplicationId=***; ApplicationSecret=***;                                       Apps=(com.microsoft.bing.halseyassistant,com.microsoft.d365.fs.mobile,com.microsoft.dynamics,com.microsoft.dynamics.inv oice,com.microsoft.dynamics.iphone.moca,com.microsoft.dynamics.iphone.moca.fieldservices,com.microsoft.dynamics.iphone. moca.sales,com.microsoft.lync2013.iphone,com.microsoft.mobile.polymer,com.microsoft.msapps,com.microsoft.msedge,com.mic rosoft.o365shdmobileapp,com.microsoft.office.excel,com.microsoft.office.outlook,com.microsoft.office.powerpoint,com.mic rosoft.office.word,com.microsoft.office365booker,com.microsoft.officelens,com.microsoft.officemobile,com.microsoft.onen ote,com.microsoft.plannermobile,com.microsoft.powerbimobile,com.microsoft.procsimo,com.microsoft.ramobile,com.microsoft .rms-sharing,com.microsoft.scmx,com.microsoft.sharepoint,com.microsoft.shiftr,com.microsoft.skydrive,com.microsoft.skyp e.teams,com.microsoft.splists,com.microsoft.stream,com.microsoft.to-do,com.microsoft.visio,com.microsoft.whiteboard,com .microsoft.workfolders,com.veradocs.ios.appstore.intune,wefwef); Assignments=(); ContactSyncBlocked=False;              CustomBrowserProtocol=; DataBackupBlocked=False; Description=some descriptive words to describe the policy in the       description. indescribable; DeviceComplianceRequired=True; DisableAppPinIfDevicePinIsSet=False;                         DisableProtectionOfManagedOutboundOpenInData=False; DisplayName=testiospolicy; Ensure=Present; ExcludedGroups=();       ExemptedAppProtocols=(Default:skype;app-settings;calshow;itms;itmss;itms-apps;itms-appss;itms-services;);               FaceIdBlocked=False; FilterOpenInToOnlyManagedApps=False; FingerprintBlocked=False;                                     Identity=T_12345678-4321-4321-4321-123456789123; ManagedBrowser=notConfigured; ManagedBrowserToOpenLinksRequired=False;  ManagedIdentity=False; MaximumPinRetries=5; MinimumPinLength=4; NotificationRestriction=allow;                         OrganizationalCredentialsRequired=False; PeriodBeforePinReset=90.00:00:00; PeriodOfflineBeforeAccessCheck=12:00:00;     PeriodOfflineBeforeWipeIsEnforced=90.00:00:00; PeriodOnlineBeforeAccessCheck=00:30:00; PinCharacterSet=numeric;         PinRequired=True; PinRequiredInsteadOfBiometricTimeout=00:30:00; PrintBlocked=False;
ProtectInboundDataFromUnknownSources=False; SaveAsBlocked=False; SimplePinBlocked=True;
TargetedAppManagementLevels=unmanaged; TenantId=***; Verbose=True
VERBOSE: [COMPUTER]:                            [[IntuneAppProtectionPolicyiOS]12345678-1234-1234-1234-123456789123]
Test-TargetResource returned False

#### Suggested solution to the issue

Adding the New-M365DSCConnection command to the start of the Test-TargetResource function appears to resolve this error and return the objects using the beta graph profile.  it's not clear to me why this should be but testing does show it to produce the correct results

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
        IntuneAppProtectionPolicyiOS 12345678-1234-1234-1234-123456789123
        {
            AllowedDataStorageLocations                    = @();
            AllowedInboundDataTransferSources              = "allApps";
            AllowedOutboundClipboardSharingExceptionLength = 0;
            AllowedOutboundClipboardSharingLevel           = "managedAppsWithPasteIn";
            AllowedOutboundDataTransferDestinations        = "allApps";
            AppActionIfDeviceComplianceRequired            = "block";
            AppActionIfIosDeviceModelNotAllowed            = "block";
            AppActionIfMaximumPinRetriesExceeded           = "block";
            AppDataEncryptionType                          = "whenDeviceLocked";
            ApplicationId                                  = $ConfigurationData.NonNodeData.ApplicationId;
            ApplicationSecret                              = New-Object System.Management.Automation.PSCredential ('ApplicationSecret', (ConvertTo-SecureString $ConfigurationData.NonNodeData.ApplicationSecret -AsPlainText -Force));
            Apps                                           = @("com.microsoft.bing.halseyassistant","com.microsoft.d365.fs.mobile","com.microsoft.dynamics","com.microsoft.dynamics.invoice","com.microsoft.dynamics.iphone.moca","com.microsoft.dynamics.iphone.moca.fieldservices","com.microsoft.dynamics.iphone.moca.sales","com.microsoft.lync2013.iphone","com.microsoft.mobile.polymer","com.microsoft.msapps","com.microsoft.msedge","com.microsoft.o365shdmobileapp","com.microsoft.office.excel","com.microsoft.office.outlook","com.microsoft.office.powerpoint","com.microsoft.office.word","com.microsoft.office365booker","com.microsoft.officelens","com.microsoft.officemobile","com.microsoft.onenote","com.microsoft.plannermobile","com.microsoft.powerbimobile","com.microsoft.procsimo","com.microsoft.ramobile","com.microsoft.rms-sharing","com.microsoft.scmx","com.microsoft.sharepoint","com.microsoft.shiftr","com.microsoft.skydrive","com.microsoft.skype.teams","com.microsoft.splists","com.microsoft.stream","com.microsoft.to-do","com.microsoft.visio","com.microsoft.whiteboard","com.microsoft.workfolders","com.veradocs.ios.appstore.intune","wefwef");
            Assignments                                    = @();
            ContactSyncBlocked                             = $False;
            CustomBrowserProtocol                          = "";
            DataBackupBlocked                              = $False;
            Description                                    = "some descriptive words to describe the policy in the description. indescribable";
            DeviceComplianceRequired                       = $True;
            DisableAppPinIfDevicePinIsSet                  = $False;
            DisableProtectionOfManagedOutboundOpenInData   = $False;
            DisplayName                                    = "testiospolicy";
            Ensure                                         = "Present";
            ExcludedGroups                                 = @();
            ExemptedAppProtocols                           = @("Default:skype;app-settings;calshow;itms;itmss;itms-apps;itms-appss;itms-services;");
            FaceIdBlocked                                  = $False;
            FilterOpenInToOnlyManagedApps                  = $False;
            FingerprintBlocked                             = $False;
            Identity                                       = "T_12345678-4321-4321-4321-123456789123";
            ManagedBrowser                                 = "notConfigured";
            ManagedBrowserToOpenLinksRequired              = $False;
            Managedidentity                                = $False;
            MaximumPinRetries                              = 5;
            MinimumPinLength                               = 4;
            NotificationRestriction                        = "allow";
            OrganizationalCredentialsRequired              = $False;
            PeriodBeforePinReset                           = "90.00:00:00";
            PeriodOfflineBeforeAccessCheck                 = "12:00:00";
            PeriodOfflineBeforeWipeIsEnforced              = "90.00:00:00";
            PeriodOnlineBeforeAccessCheck                  = "00:30:00";
            PinCharacterSet                                = "numeric";
            PinRequired                                    = $True;
            PinRequiredInsteadOfBiometricTimeout           = "00:30:00";
            PrintBlocked                                   = $False;
            ProtectInboundDataFromUnknownSources           = $False;
            SaveAsBlocked                                  = $False;
            SimplePinBlocked                               = $True;
            TargetedAppManagementLevels                    = "unmanaged";
            TenantId                                       = $ConfigurationData.NonNodeData.TenantId;
        }
```

#### Version of the DSC module that was used ('dev' if using current dev branch)
dev

### PR DESCRIPTION
**Pull request by menswearUK**
added New-M365DSCConnection command to address #2942 - added formatting to current/target values output

#### Pull Request (PR) description

This PR adds a line:
$ConnectionMode = New-M365DSCConnection -Workload 'MicrosoftGraph' -ProfileName beta `
    -InboundParameters $PSBoundParameters

to the Test-TargetResource function of IntuneAppProtectionPolicyiOS

It prevents the Get-TargetResource function returning objects with missing values.  It's not clear to me why the same line in get-TargetResource doesn't work correctly but without this line added here the object returned from the MS graph command uses the v1.0 profile and is missing a number of values which are required for the returned array.  This leads to failed comparisons with config documents even when the settings are correct.

The error and the fix can be seen in a verbose output of start-DSCConfiguration or in test-DSCConfiguration

#### This Pull Request (PR) fixes the following issues

Fixes #2942

### CODE REVIEW

**Comment by ykuijs** [comment]
Even though this solution works, it is more like a workaround. I rather try to solve the underlying issue before implementing a workaround.

**Comment by menswearUK** [comment]
I thought it was easiest to update this PR with the places I added code into New-M365DscConnection - since the PR is on hold and, once we reach a solution, I can amend further as necessary.
In addition to my initial suggestion I also added a check that the New-M365DscConnection was for MSGraph before applying the switch as it could be used for multiple connections.
In my testing there was no requirement to add further verbose settings changes if the command is run within New-M365DscConnection but I can add those in too if they are required

**Comment by menswearUK** [comment]
@ykuijs  - I've updated this PR with the fixes discussed in the issue chat - the only thing I've not done is added in the changes to the verbose output (it only seems to be an issue when the profile switch is in the configuration, it doesn't seem to be required in this function although I can add those lines in if you'd prefer).

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # Change log for Microsoft365DSC
 
 # UNRELEASED
+* New-M365DSCConnection in M365DSCUtil.psm1
```
Comment (ykuijs) [inline CHANGELOG.md:4]: Can you please move this item under the MISC section (currently on line 14)

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.psm1 [hunk 2]
```diff
@@ -1080,8 +1080,8 @@ function Test-TargetResource
 
     $CurrentValues = Get-TargetResource @PSBoundParameters
 
-    Write-Verbose -Message "Current Values: $(Convert-M365DscHashtableToString -Hashtable $CurrentValues)"
-    Write-Verbose -Message "Target Values: $(Convert-M365DscHashtableToString -Hashtable $PSBoundParameters)"
+    Write-Verbose -Message "Current Values: $((Convert-M365DscHashtableToString -Hashtable $CurrentValues) -replace ';', "`r`n")"
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.psm1:1083]: Can you leave this as it was? I like the change, but now it is only implemented for this resource and not for all other resources. I rather change the Convert-M365DscHashtableToString function to do this, so all resources replace the semi colon with an enter.

Please create an issue for this so we can assess the impact and implement separately.

File: Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1 [hunk 3]
```diff
@@ -1575,6 +1575,13 @@ function New-M365DSCConnection
                 -Credential $InboundParameters.Credential `
                 -SkipModuleReload $Global:CurrentModeIsExport `
                 -ProfileName $ProfileName
+
+            #2942 - check correct profile applied
+            if (($ProfileName -ne $global:MSCloudLoginConnectionProfile.MicrosoftGraph.ProfileName) -and ($Workload -eq 'MicrosoftGraph') )
+            {
+                select-mgprofile -Name $profilename
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1:1582]: Could you update the cmdlet to match the default PascalCase format: Select-MgProfile

And update the variable name to match the earlier used PascalCasing: $ProfileName

Both are true for all instances you have added.

**Review by ykuijs — CHANGES_REQUESTED** [review]
Thanks for this update. Just a few minor comments.

**Comment by menswearUK** [comment]
@ykuijs No worries, I've corrected those values now.

**Comment by ykuijs** [comment]
@menswearUK, thanks for that. But what about the other two comments?

**Review by ykuijs — APPROVED** [review]
LGTM

**Comment by ykuijs** [comment]
@menswearUK, wanted to merge it, so I could include it in this weeks release. But unfortunately I was unable to. Could you update the PR with the Dev branch?

**Comment by menswearUK** [comment]
@ykuijs - apologies for this, I was attempting to clear the conflicts and messed it up - I ended up clearing the commits and syncing and it closed the PR with no changes.  I've created a new branch with the updates which has no conflicts and I've put in  a new PR for that.

**Pull request closed without merging**